### PR TITLE
Fix On RefundEscrow, add isState function #1461

### LINKED ISSUE #1461: On RefundEscrow, add isState function
We recommend replacing all ​ _state​ checks such as require(_state == State.Active);​ with a modifier. (e.g. ​ isState(State.Active)​ )

This issue was reported in the [LevelK](https://www.levelk.io/) audit.

### PR DESCRIPTION
**Pull request by obernardovieira**
Fixes #1461 

> We recommend replacing all ​ _state​ checks such as require(_state == State.Active);​ with a modifier. (e.g. ​ isState(State.Active)​ )
> This issue was reported in the LevelK audit.

So I just added the modifier and updated the code has said in issue. There is no new test and I did run test and coverage to make sure everything was alright.

Before submitting, please make sure that you have:
  - [x] reviewed the OpenZeppelin [Contributor Guidelines](https://github.com/OpenZeppelin/openzeppelin-solidity/blob/master/CONTRIBUTING.md),
  - [ ] added tests where applicable to test new functionality,
  - [x] made sure that your contracts are well-documented
  - [x] run the JS/Solidity linters and fixed any issues (`npm run lint:fix`).

### CODE REVIEW

**Comment by obernardovieira** [comment]
I've seen the following on travis build
```bash
,/home/travis/build/OpenZeppelin/openzeppelin-solidity/contracts/payment/escrow
/RefundEscrow.sol:45:20: Warning: This declaration shadows an existing declaration.
  modifier isState(State state) {
                   ^---------^
```
is it because there's a method named `state()` ?
I would like to fix the warning. Thanks.

File: contracts/payment/escrow/RefundEscrow.sol [hunk 1]
```diff
@@ -39,6 +39,14 @@ contract RefundEscrow is ConditionalEscrow {
     return _state;
   }
 
+  /**
+   * @dev Throws if called by any state other than the expected one.
+   */
+  modifier isState(State state) {
```
Comment (obernardovieira) [inline contracts/payment/escrow/RefundEscrow.sol]:
```suggestion
  modifier isState(State expected) {
```

File: contracts/payment/escrow/RefundEscrow.sol [hunk 2]
```diff
@@ -39,6 +39,14 @@ contract RefundEscrow is ConditionalEscrow {
...
+    require(_state == state);
```
Comment (obernardovieira) [inline contracts/payment/escrow/RefundEscrow.sol]:
```suggestion
    require(_state == expected);
```

**Comment by obernardovieira** [comment]
Hey guys,
It's been a while. I fixed a conflict it had. Looks like validations are ok. Is it possible to merge?

File: contracts/payment/escrow/RefundEscrow.sol [hunk 3]
```diff
@@ -31,6 +31,14 @@ contract RefundEscrow is ConditionalEscrow {
         _beneficiary = beneficiary;
         _state = State.Active;
     }
+    
+    /**
+     * @dev Throws if called by any state other than the expected one.
+     */
+    modifier isState(State expectedState) {
```
Comment (nventuro) [inline contracts/payment/escrow/RefundEscrow.sol]:
```suggestion
    modifier onlyInState(State expectedState) {
```

File: contracts/payment/escrow/RefundEscrow.sol [hunk 4]
```diff
@@ -50,35 +58,31 @@ contract RefundEscrow is ConditionalEscrow {
      * @dev Stores funds that may later be refunded.
      * @param refundee The address funds will be sent to if a refund occurs.
      */
-    function deposit(address refundee) public payable {
-        require(_state == State.Active);
+    function deposit(address refundee) public payable isState(State.Active) {
```
Comment (nventuro) [inline contracts/payment/escrow/RefundEscrow.sol]:
```suggestion
    function deposit(address refundee) public payable onlyInState(State.Active) {
```

File: contracts/payment/escrow/RefundEscrow.sol [hunk 5]
```diff
@@ -50,35 +58,31 @@ contract RefundEscrow is ConditionalEscrow {
...
         super.deposit(refundee);
     }
 
     /**
      * @dev Allows for the beneficiary to withdraw their funds, rejecting
      * further deposits.
      */
-    function close() public onlyPrimary {
-        require(_state == State.Active);
+    function close() public onlyPrimary isState(State.Active) {
```
Comment (nventuro) [inline contracts/payment/escrow/RefundEscrow.sol]:
```suggestion
    function close() public onlyPrimary onlyInState(State.Active) {
```

File: contracts/payment/escrow/RefundEscrow.sol [hunk 6]
```diff
@@ -50,35 +58,31 @@ contract RefundEscrow is ConditionalEscrow {
...
         _state = State.Closed;
         emit RefundsClosed();
     }
 
     /**
      * @dev Allows for refunds to take place, rejecting further deposits.
      */
-    function enableRefunds() public onlyPrimary {
-        require(_state == State.Active);
+    function enableRefunds() public onlyPrimary isState(State.Active) {
```
Comment (nventuro) [inline contracts/payment/escrow/RefundEscrow.sol]:
```suggestion
    function enableRefunds() public onlyPrimary onlyInState(State.Active) {
```

File: contracts/payment/escrow/RefundEscrow.sol [hunk 7]
```diff
@@ -50,35 +58,31 @@ contract RefundEscrow is ConditionalEscrow {
...
         _state = State.Refunding;
         emit RefundsEnabled();
     }
 
     /**
      * @dev Withdraws the beneficiary's funds.
      */
-    function beneficiaryWithdraw() public {
-        require(_state == State.Closed);
+    function beneficiaryWithdraw() public isState(State.Closed) {
```
Comment (nventuro) [inline contracts/payment/escrow/RefundEscrow.sol]:
```suggestion
    function beneficiaryWithdraw() public onlyInState(State.Closed) {
```

**Comment by obernardovieira** [comment]
Thanks, @nventuro 
it looks good to me :+1: should I applied the suggestions.

**Comment by frangio** [comment]
@obernardovieira We have been thinking about this and have decided not to make this change. The main reason is that our coverage analysis tool does not properly work with modifiers. Apologies for the unmerged work!

**Pull request closed without merging**